### PR TITLE
Created get/update wallet

### PR DESCRIPTION
**Pull request by prakashchoudhary07**

### CODE REVIEW

File: routes/index.js [hunk 1]
```diff
@@ -11,5 +11,6 @@ app.use('/pullrequests', require('./pullrequests.js'))
 app.use('/contributions', require('./contributions'))
 app.use('/badges', require('./badges.js'))
 app.use('/auctions', require('./auctions.js'))
+app.use('/wallets', require('./wallets.js'))
```
Comment (swarajpure) [inline routes/index.js:14]: `wallet` or `wallet`?

File: controllers/walletsController.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+const { fetchWallet } = require('../models/wallets')
+
+const ERROR_MESSAGE = 'Something went wrong. Please try again or contact admin'
+
+/**
+ * Get the  contributions of the user
```
Comment (sumitd94) [inline controllers/walletsController.js:6]: `Contributions`?

File: controllers/walletsController.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * @param {Object} req - Express request object
+ * @param {Object} res - Express response object
+ */
+
+const getUserWallet = async (req, res) => {
+  try {
+    const { id: userId } = req.userData
+    const walletData = await fetchWallet(userId)
+    return res.status(200).json({
```
Comment (sumitd94) [inline controllers/walletsController.js:15]: NIT: By default, it returns 200, we can return this like `res.json({})`

File: models/wallets.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    const walletData = {
+      userId,
+      isActive: true,
+      currency: {}
+    }
+    const wallet = await walletModel.add(walletData)
+    return {
+      id: wallet.id,
+      data: walletData
+    }
```
Comment (sumitd94) [inline models/wallets.js:36]: NIT: we can destructure here and make it like this
```javascript
const { id } = await walletModel.add(walletData)
return {
id,
data: walletData
}
```

File: models/wallets.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  }
+}
+
+/**
+ * Update currecny wallet for user
+ * @return {Promise<walletModel|object>}
+ */
+const updateWallet = async (userId, currency) => {
+  try {
+    let userWallet = await fetchWallet(userId)
```
Comment (sumitd94) [inline models/wallets.js:49]: NIT: If from `userWallet` we only need `id` then we can destructure only `id` here

**Review by sumitd94 — APPROVED** [review]
Few NIT Comments, This was perfect FTL Work 🔥

File: routes/wallets.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
+const express = require('express')
+const router = express.Router()
+const walletController = require('../controllers/walletsController')
+const authenticate = require('../middlewares/authenticate')
+
+/**
+ * @swagger
+ * /wallet/:
+ *   get:
+ *     summary: Gets the user walelt details
```
Comment (ankushdharkar) [inline routes/wallets.js:10]: *wallet

File: routes/index.js (hunk 1 above)
Comment (ankushdharkar) [reply routes/index.js:14]: I'd go with `wallet`

File: models/wallets.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * Fetches the data from user wallet
+ * @return {Promise<walletModel|object>}
+ */
+const fetchWallet = async (userId) => {
+  try {
+    const walletDocs = await walletModel.where('userId', '==', userId).limit(1).get()
+    if ((!walletDocs.empty) && walletDocs.docs[0]) {
+      return { id: walletDocs.docs[0].id, ...walletDocs.docs[0].data() }
```
Comment (ankushdharkar) [inline models/wallets.js:12]: ```const userWallet = walletDocs.docs[0];```

File: models/wallets.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+const walletModel = firestore.collection('wallet')
+
+/**
+ * Fetches the data from user wallet
+ * @return {Promise<walletModel|object>}
+ */
+const fetchWallet = async (userId) => {
+  try {
+    const walletDocs = await walletModel.where('userId', '==', userId).limit(1).get()
+    const userWallet = walletDocs.docs[0]
```
Comment (ankushdharkar) [inline models/wallets.js:11]: NIT: `const [ userWallet ] = walletDocs.docs;`

File: models/wallets.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    if ((!walletDocs.empty) && userWallet) {
```
Comment (ankushdharkar) [inline models/wallets.js:12]: Why do we need to check for `walletDocs.empty`?

File: models/wallets.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * Update currecny wallet for user
+ * @return {Promise<walletModel|object>}
+ */
+const updateWallet = async (userId, currency) => {
+  try {
+    let userWallet = await fetchWallet(userId)
+    if (!userWallet.id) {
+      await createWallet(userId)
+      userWallet = await fetchWallet(userId)
```
Comment (ankushdharkar) [inline models/wallets.js:53]:
```js
 userWallet = await createWallet(userId)
 ```

File: models/wallets.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    }
+    const firestoreKeyName = {}
```
Comment (ankushdharkar) [inline models/wallets.js:55]: firestoreKeysObj

File: controllers/walletsController.js [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+ * Get the  wallet details of user
+ * @param {Object} req - Express request object
+ * @param {Object} res - Express response object
+ */
+
+const getUserWallet = async (req, res) => {
+  try {
+    const { id: userId } = req.userData
+    const walletData = await fetchWallet(userId)
+    return res.status.json({
```
Comment (ankurnarkhede) [inline controllers/walletsController.js:15]: return res.json(...

**Review by ankushdharkar — APPROVED** [review]

File: models/wallets.js [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * Create new wallet for user
+ * @return {Promise<walletModel|object>}
+ */
+const createWallet = async (userId) => {
+  try {
+    const walletData = {
+      userId,
+      isActive: true,
+      currency: {}
```
Comment (ankurnarkhede) [inline models/wallets.js:31]: I am assuming the `currency` object will contain values like
```
"gold": {// gold coins details}
"neelam": // details
...etc
```

Or will it be different?
```

**Review by ankurnarkhede — CHANGES_REQUESTED** [review]

File: models/wallets.js [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    }
+    const { id } = await walletModel.add(walletData)
```
Comment (correaswebert) [inline models/wallets.js:33]: user-wallet is a one-one relation
so keeping the userId == walletId would be easier right?

File: models/wallets.js [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    }
+    const firestoreKeyName = {}
+    for (const key in currency) {
+      firestoreKeyName[`currency.${key}`] = currency[`${key}`]
+    }
+    const walletRef = walletModel.doc(userWallet.id)
+    const res = await walletRef.update({
+      ...firestoreKeyName
+    })
+    if (res) {
```
Comment (correaswebert) [inline models/wallets.js:62]: return Boolean(res);

**Pull request merged**